Created tests for USGS Spectral Library Version 7 convert function

bdegley4789 opened the pull request:
This doesn't need to get accepted right now. I still need to get all the tests to pass, I had said I would do a pr either today or yesterday so you could review what I have been working on before our meeting on Friday at 8am. 
It looks like this gets some errors with test_classify_image. I'm guessing this is because I changed the envi .hdr and .sli files from the spectral v06 to v07 files

Commit:
splib07

Commit:
Merge pull request #1 from capstone-coal/master

update branch

Commit:
spectral 7 converter start

Commit:
Splib 07 test one file

Commit:
Finished Spectral Version 7 convert

Commit:
Finish client changes to spectral convert function

Commit:
Condensed to one script

Commit:
Fix oswalk and other changes

Commit:
Remove duplicate spectra

Commit:
remove aster files

Commit:
Remove duplicate spectra

Commit:
Merge branch 'master' of https://github.com/bdegley4789/pycoal

Commit:
Change to unknown

Commit:
Fix spelling

Commit:
Move convert function to mineral.py

Commit:
Fix band num

Commit:
Merge pull request #2 from capstone-coal/master

Fix broken location of GDAL CURRENT to fixed 2.2.4 in Dockerfile

Commit:
Fix reload error travis ci

Commit:
update reload to py3

Commit:
remove unused import

Commit:
fix mineral import

Commit:
remove sys path import

Commit:
removed bad import statements

Commit:
Merge branch 'master' of https://github.com/bdegley4789/pycoal

Commit:
Change spectra names to v7

Commit:
Worked on testing

lewismc commented on examples/example_mineral.py:
```diff
@@ -61,7 +61,7 @@
 
 input_filename = 'avng.jpl.nasa.gov/AVNG_2015_data_distribution/L2/ang20150420t182050_rfl_v1e/ang20150420t182050_corr_v1e_img.hdr'
-library_filename='../pycoal/tests/s06av95a_envi.hdr'
+library_filename='../pycoal/tests/s07_AV95_envi.hdr'
```
You need to make sure that there are no regressions what-so-ever. We are NOT dropping support for USGS v6. We are _**adding**_ functionality to use v7... this means that all v6 code should not be touched.
The solution here is to either

1. introduce a **_brand new example file_** for v7 mineral classification, or 
1. introduce a new CLI flag which determines the spectral library at runtime based upon the program input.

lewismc commented on examples/example_mineral.py:
```diff
@@ -120,7 +120,7 @@ def main(argv=None):
         parser = ArgumentParser(description=program_license, formatter_class=RawDescriptionHelpFormatter)
         parser.add_argument("-i", "--image", dest="image", default=input_filename, help="Input file to be processed [default: ang20150420t182050_corr_v1e_img.hdr]")
-        parser.add_argument("-s", "--slib", dest="slib", default=library_filename, help="Spectral Library filename [default: s06av95a_envi.hdr]")
+        parser.add_argument("-s", "--slib", dest="slib", default=library_filename, help="Spectral Library filename [default: s07_AV95_envi.hdr]")
```
Same here...

lewismc commented on examples/example_spectral07.py:
```diff
@@ -0,0 +1,49 @@
+
+Dependencies
+`USGS Spectral Library Version 7 <https://speclab.cr.usgs.gov/spectral-lib.html>`_
+must be downloaded to the examples directory
```
> must be downloaded to the examples directory

should be changed to 

> must be downloaded and unzipped to the examples directory

lewismc commented on examples/example_spectral07.py:
```diff
@@ -0,0 +1,49 @@
+
+@author:     COAL Developers
+
+@copyright:  2018 COAL Developers. All rights reserved.
```
change to ```2017-2018```

lewismc commented on examples/example_spectral07.py:
```diff
@@ -0,0 +1,49 @@
+@contact:    coal-capstone@googlegroups.com
+'''
+
+#!/usr/bin/python
```
This always goes at the top of the file.

lewismc commented on examples/example_spectral07.py:
```diff
@@ -0,0 +1,49 @@
+usgs_convolved = mineral.FullSpectralLibrary7Convert()
+usgs_convolved.convert('usgs_splib07')
+
+
```
remove extra line

lewismc commented on pycoal/mineral.py:
```diff
@@ -14,12 +14,18 @@
 
+import sys
+import os
+import inspect
```
Remove unused import

lewismc commented on pycoal/mineral.py:
```diff
@@ -14,12 +14,18 @@
 import time
+import fnmatch
+import shutil
+import mmap
```
Remove unused import

lewismc commented on pycoal/mineral.py:
Writing all of these values still seems way too tightly tied with some very specific file. It is not correct and needs to be much more flexible.

lewismc commented on pycoal/mineral.py:
```diff
@@ -362,3 +368,184 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
 
         library.save(hdr_file)
+
+class SpectralToAsterConversion:
```
Change this Class name to something more descriptive.

lewismc commented on pycoal/mining.py:
```diff
@@ -21,9 +21,9 @@
-proxy_class_names = [u'Schwertmannite BZ93-1 s06av95a=b',
-                   u'Renyolds_TnlSldgWet SM93-15w s06av95a=a',
-                   u'Renyolds_Tnl_Sludge SM93-15 s06av95a=a']
+proxy_class_names = [u'Schwertmannite BZ93-1         BECKb AREF',
```
Again, this cannot change... we need to maintain backwards compatibility and **add** functionality not take away.

lewismc commented on pycoal/tests/test.py:
```diff
@@ -26,7 +26,7 @@ def _remove_files(listOfFileNames):
 
 # file names of USGS Digital Spectral Library 06 in ENVI format
-libraryFilenames = ["s06av95a_envi.hdr", "s06av95a_envi.sli"]
+libraryFilenames = ["s07_AV95_envi.hdr", "s07_AV95_envi.sli"]
```
... and again. You need to add new entries not take away from what we already have.

lewismc commented on pycoal/tests/test.py:
```diff
@@ -26,7 +26,7 @@ def _remove_files(listOfFileNames):
+libraryFilenames = ["s07_AV95_envi.hdr", "s07_AV95_envi.sli"]
 
 # set up test module before running tests
 def setup_module(module):
```
You are not using any of the test files you've introduced. You either need to use them or remove them.

lewismc reviewed (changes requested):
@bdegley4789 please look at my comments. Thank you for writing unit tests. Lets talk tomorrow.

lewismc commented:
@bdegley4789 I would also like to remind you to always test your code before you submit a pull request. It is VERY simple and takes no longer than about 2 seconds to type and around a minute or so to execute.
https://github.com/capstone-coal/pycoal#tests
Thanks

lewismc commented:
@bdegley4789 any updates here?

bdegley4789 commented:
@lewismc I can’t really figure why my tests aren’t passing. I re-downloaded pycoal from the master. I moved the files I changed over to the new pycoal I downloaded and now everything passes. I must have changed something in a directory I’m unaware of. I’d like to figure out what is going on first, if I can't figure it out, I'll just push this new version to my fork and update this pr tomorrow or after I finish work today. 

This is what I get when I move all the files I changed over to a new pycoal directory.

```
Using /Users/bryceegley/anaconda3/lib/python3.6/site-packages
Finished processing dependencies for pycoal==0.5.2
Bryces-MacBook-Air:pycoal bryceegley$ nosetests
0...10...20...30...40...50...60...70...80...90...100 - done.
0...10...20...30...40...50...60...70...80...90...100 - done.
...........
----------------------------------------------------------------------
Ran 11 tests in 9.731s

OK
Bryces-MacBook-Air:pycoal bryceegley$ 
```

On my current fork I still get these 2 failures and 1 error

```
ERROR: mineral_test.test_classify_image_subset
Traceback (most recent call last):
  File "/Users/bryceegley/anaconda3/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/bryceegley/Desktop/temp/pycoal/pycoal/tests/mineral_test.py", line 135, in test_classify_image_subset
    mc = mineral.MineralClassification(libraryFilenames[0], class_names=mining.proxy_class_names)
  File "/Users/bryceegley/Desktop/temp/pycoal/pycoal/mineral.py", line 60, in __init__
    self.library = self.subset_spectral_library(self.library, class_names)
  File "/Users/bryceegley/Desktop/temp/pycoal/pycoal/mineral.py", line 309, in subset_spectral_library
    old_index = spectral_library.names.index(class_name)
ValueError: 'Schwertmannite BZ93-1 s06av95a=b' is not in list


FAIL: mineral_test.test_classify_image
Traceback (most recent call last):
  File "/Users/bryceegley/anaconda3/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/bryceegley/Desktop/temp/pycoal/pycoal/tests/mineral_test.py", line 69, in test_classify_image
    assert expected.metadata.get(u'class names') == actual.metadata.get(u'class names')
AssertionError: 


FAIL: mineral_test.test_classify_image_in_memory
Traceback (most recent call last):
  File "/Users/bryceegley/anaconda3/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/bryceegley/Desktop/temp/pycoal/pycoal/tests/mineral_test.py", line 94, in test_classify_image_in_memory
    assert numpy.array_equal(expected.asarray(), actual.asarray())
AssertionError: 
```

bdegley4789 commented:
I also finished running the mineral classification using the Spectral Version 7 envi .hdr and .sli on the default example and staged it here
 https://drive.google.com/drive/folders/1LTJv5H6_YxMDz5fff074MUj59QO9mwvH

Based on the three mineral classifications I’ve done with Spectral v7, everything seems to perform normally using the Spectral Version 7 envi .hdr and .sli files

lewismc commented:
Bryce, look at the stack trace...

ValueError: 'Schwertmannite BZ93-1 s06av95a=b' is not in list

lewismc commented:
You’ve also neglected to address any of my comments... come on.

bdegley4789 commented:
I'm working on implementing the comments you made, I was just giving an update on where I was with that. 

I changed mining.py back to the names of Spectral Version 6 Library so I thought 'Schwertmannite BZ93-1 s06av95a=b' would be in the list. 
I did that to keep support for Spectral v6 library like we discussed.

I was thinking about doing multiple examples one for Spectral v7 and one for Spectral v6. But mining.py requires the spectra names so I think the only option
of doing this is through the other method you mentioned using the CLI

lewismc commented:
> I was thinking about doing multiple examples one for Spectral v7 and one for Spectral v6. But mining.py requires the spectra names so I think the only option
of doing this is through the other method you mentioned using the CLI

Why don't you just do the following

```
# classes identified as proxies for coal mining classification using USGSv6
proxy_class_names_usgsv6 = [u'Schwertmannite BZ93-1 s06av95a=b',
                   u'Renyolds_TnlSldgWet SM93-15w s06av95a=a',
u'Renyolds_Tnl_Sludge SM93-15 s06av95a=a']

# classes identified as proxies for coal mining classification using USGSv7
proxy_class_names_usgsv7 = [u'Schwertmannite BZ93-1 BECKb AREF',
                   u'Renyolds_TnlSldgWet SM93-15w BECKa AREF',
u'Renyolds_Tnl_Sludge SM93-15 BECKa AREF']
```

Then use the ```proxy_class_names_usgsv6``` and ```proxy_class_names_usgsv7``` depending on whether USGSv6 or USGSv7 classification is required?

lewismc commented:
Take a look at the comments... you've not addressed them. If this is not done by Thursday morning I will just do it myself.

Commit:
Allow cli to work with spectral v7 and v6

bdegley4789 commented:
@lewismc Ok, I’m just using two different examples scripts for mineral classification to run USGS Spectral v7 and v6 libraries. Mining.py now has a new argument in the CLI which can be passed as a number either 6 or 7. Or you can just change the number in example_mining.py

I tested running this with 6 and 7 and it works. But it must be the same version number as you did in the mineral classification to get the classified file. If the classified image was created with USGS Spectral Version 6, you can’t do a mining classification with version 7 since the spectra names are different.

```
Bryces-MacBook-Air:examples bryceegley$ python example_mining.py
Saving /Users/bryceegley/Desktop/pycoal/examples/ang20150420t182050_corr_v1e_img_class_mining.img
Bryces-MacBook-Air:examples bryceegley$
```

I fixed errors in nose tests, now everything passes and I made all the requested changes you had
```
Bryces-MacBook-Air:pycoal bryceegley$ nosetests
0...10...20...30...40...50...60...70...80...90...100 - done.
0...10...20...30...40...50...60...70...80...90...100 - done.
............
----------------------------------------------------------------------
Ran 12 tests in 9.768s

OK
Bryces-MacBook-Air:pycoal bryceegley$ 
```

bdegley4789 commented:
I don't know why travis-ci keeps saying `FileNotFoundError: [Errno 2] No such file or directory: 'pycoal/tests'`

lewismc commented:
@bdegley4789 what are all of the files 
```
ASCIIdata/ASCIIdata_splib07a/splib07a_Bandpass_(FWHM)_ASDNG_High-Res_NextGen.txt
```
If you re not using them remove them.

lewismc commented:
> I don't know why travis-ci keeps saying FileNotFoundError: [Errno 2] No such file or directory: 'pycoal/tests'

Look at the error traces in the TravisCI logs

```
======================================================================
ERROR: test suite for <module 'environment_test' from '/home/travis/build/capstone-coal/pycoal/pycoal/tests/environment_test.py'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/home/travis/virtualenv/python3.5.5/lib/python3.5/site-packages/nose/util.py", line 469, in try_run
    return func(obj)
  File "/home/travis/build/capstone-coal/pycoal/pycoal/tests/test.py", line 47, in setup_module
    ftp.retrbinary('RETR %s' % f, lib_f.write)
  File "/opt/python/3.5.5/lib/python3.5/ftplib.py", line 443, in retrbinary
    with self.transfercmd(cmd, rest) as conn:
  File "/opt/python/3.5.5/lib/python3.5/ftplib.py", line 400, in transfercmd
    return self.ntransfercmd(cmd, rest)[0]
  File "/opt/python/3.5.5/lib/python3.5/ftplib.py", line 362, in ntransfercmd
    source_address=self.source_address)
  File "/opt/python/3.5.5/lib/python3.5/socket.py", line 712, in create_connection
    raise err
  File "/opt/python/3.5.5/lib/python3.5/socket.py", line 703, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out
```
There is a timeout when executing the FTP file retrieval in  ```  File "/home/travis/build/capstone-coal/pycoal/pycoal/tests/test.py", line 47, in setup_module```

lewismc commented:
How large are these file? If they are not large then we should just cache them locally within the tests directory as a resource.

Commit:
Fix ftp url

bdegley4789 commented:
Those files in `pycoal/pycoal/tests/usgs_splib07/ASCIIdata/` are for testing. I did the same thing as the ASTER tests. Take a small subset of the files in the spectral library and use them to run my conversion function on to verify that it generates a .sli and .hdr file. I could probably remove some of them but to test my conversion functions we should have some files in there. It is 221 KB for the files I included for testing. The ASTER test files are about half that 106KB so I could remove some if that is too much space.

lewismc commented:
Can you please squash your commits. This PR is very messy.

> Those files in pycoal/pycoal/tests/usgs_splib07/ASCIIdata/ are for testing.

Unless they are being used in unit tests (which per your PR they are not) then they should be removed from the PR.

Commit:
Remove ftp calls, fix overwrite error in nosetests

Commit:
Remove Spectral 7 library

Commit:
Merge pull request #3 from capstone-coal/master

remove s06_envi files from gitignore

Commit:
Repush with gitignore updated

bdegley4789 commented:
@lewismc I fixed the errors with the ftp statements we discussed, I pushed the convolved envi .hdr and .sli files and I fixed another error I found when running nosetests. It looks like all the checks now pass. I think this pull request is good to be accepted and fulfills the work needed on issue #105 . In the next two days I'll add documentation for this issue to the website. Let me know what you think

lewismc commented:
Excellent work on this issue @bdegley4789 it's been a long time coming and a lot of back and forth these last few months... but it's now ready to commit to master branch. Excellent work, thank you for the persistence.
